pages/data-providers/[data-provider-id]/settings.jsx

```diff
@@ -0,0 +1,105 @@
+import React, { useState, useEffect } from 'react'
+import { classNames } from '@oacore/design/lib/utils'
+
+import styles from './settings.css'
+
+import { Card, TextField, Button } from 'design'
+import { withGlobalStore } from 'store'
+
+const SettignsGroup = ({
+  children,
+  className,
+  tag,
+  onSubmit,
+  ...formProps
+}) => {
+  const [isChanged, setChanged] = useState(false)
+  const [isSubmited, setSubmitted] = useState(false)
+
+  const handleChange = () => setChanged(true)
+
+  const handleSubmit = event => {
+    setSubmitted(true)
+    if (onSubmit) onSubmit(event)
+  }
+
+  useEffect(() => {
+    setChanged(false)
+    setSubmitted(false)
+  }, [children])
```

Comment:
Relying on children may as dangerous as relaying on only other non-primitive value. However, it works and it makes me wondering.

What do you think?

Reply:
I consider this as an antipattern. The only reason why it works is just because parent component doesn't get rerendered. This effect gets called always when parent component is rerendered.

See this simple [example](https://codesandbox.io/s/icy-pine-rwzu0?fontsize=14&hidenavigation=1&theme=dark)

Reply:
I agree. That makes the component unpredictable what is bad even if it works. Do you have any suggestions?

---

I really like your example 👍

Reply:
What about moving it to handleSubmit? If onSubmit doesn't throw any exception set it back to false.

Reply:
I consider awaiting in event listeners an antipattern too. It's as unpredictable as comparing children. There must be a better way to handle that.

Reply:
Why is it unpredictable?

Reply:
Because when you have a central state management (Model, Store, MobX in our case) you expect every component is a pure function calculated from the application state.

Awaiting breaks the purity as well as comparing children. It stores some state in the closure and checks it later. Although, store update must make the components to update, only.

I feel okay with handling onChange event in local state because it's a simple enhancement — indicating changes. If the button is always shown, nothing would change. However, processing requests it different.